src/estimation/alignment_properties.rs

```diff
@@ -83,42 +85,104 @@ impl AlignmentProperties {
 
-            if is_regular
-                && !record.is_mate_unmapped()
-                && record.is_first_in_template()
-                && record.tid() == record.mtid()
-                && record.mapq() > 0
+            // Records to skip after updating max_cigar_ops_len, BUT without incrementing the
+            // counter (to keep looking for 10000 useful records for the estimation)
+            if !record.is_paired()
+                || !record.is_first_in_template()
+                || !(record.tid() == record.mtid())
+                || record.is_mate_unmapped()
             {
+                skipped += 1;
+                continue;
+            }
+
+            if !is_regular {
+                n_not_useable += 1;
+                if has_soft_clip {
+                    n_soft_clip += 1;
+                }
+            } else {
                 // record insert size
                 tlens.push(record.insert_size().abs() as f64);
             }
 
             i += 1;
         }
 
-        let upper = tlens.percentile(95);
-        let lower = tlens.percentile(5);
-        let mut valid = tlens
-            .into_iter()
-            .filter(|l| *l <= upper && *l >= lower)
-            .collect_vec();
-        properties.insert_size.mean = valid.median();
-        properties.insert_size.sd = valid.iter().std_dev();
-
-        Ok(properties)
+        if tlens.len() == 0 {
+            eprint!(
+                "\nFound no records to use for estimating the insert size. Will assume\n\
+                    single end sequencing data and calculate deletion probabilities without\n\
+                    considering the insert size.\n\
+                    \n\
+                    If your data should be paired end, please consider manually providing\n\
+                    --alignment-properties, e.g. computed with `samtools stats`. Also,\n\
+                    the following counts of unusable records might indicate a source of\n\
+                    this problem:\n\n\
+                    - I, D, S or H CIGAR operation: {nu}\n\
+                    - S CIGAR (soft clip, e.g. due to UMIs or adapters): {sc}\n\
+                    \n\
+                    In addition, {nr} records were skipped in the estimation for one\n\
+                    of the following reasons:\n\
+                       - not paired\n\
+                       - not the first segment with regard to the template sequence\n\
+                       - mapping quality of 0\n\
+                       - marked as a duplicate\n\
+                       - mate mapped to different template (e.g. different chromosome)\n\
+                       - failed some quality check according to the 512 SAM flag\n\
+                       - mate unmapped\n\
+                       - record unmapped\n",
+                nu = n_not_useable,
+                sc = n_soft_clip,
+                nr = skipped
+            );
+            properties.insert_size.mean = std::f64::NAN;
```

Comment:
The idiomatic way is to use Option<f64> instead, and set it so `None` here. This enforces all code that uses the property object to specifically handle the None case, instead of accidentally calculating stuff using the NaN values.

Reply:
Just to check that I understand correctly, before implementing this... I would wrap the `InsertSize` into an option and then make it a `None` here and a `Some(InsertSize)` everywhere else?

So for the general deifition:
```
pub struct AlignmentProperties {
    insert_size: Option<InsertSize>,
    pub(crate) max_del_cigar_len: u32,
    pub(crate) max_ins_cigar_len: u32,
    pub(crate) frac_max_softclip: f64,
}
```

And here:
```
properties.insert_size = None;
```

The rest of the property object may well be usable, so we would not want to set the whole `properties` to `None`, right?

Reply:
Yes, exactly.